swerve/SwerveSubsystems.py

```diff
@@ -0,0 +1,220 @@
+    SwerveConstants.bl_drive_id,
+    SwerveConstants.bl_turn_id,
+    abs_encoder_offset_rad=SwerveConstants.bl_abs_encoder_offset_rad,
+    chassis_angular_offset=SwerveConstants.bl_chassis_angular_offset,
+  )
+  back_right = SwerveModule(
+    SwerveConstants.br_drive_id,
+    SwerveConstants.br_turn_id,
+    abs_encoder_offset_rad=SwerveConstants.br_abs_encoder_offset_rad,
+    chassis_angular_offset=SwerveConstants.br_chassis_angular_offset,
+  )
+
+  odometer = SwerveDrive4Odometry(
+    SwerveConstants.kDriveKinematics,
+    Rotation2d(),
+    (
+      front_left.getPosition(),
+      front_right.getPosition(),
+      back_left.getPosition(),
+      back_right.getPosition(),
+    ),
+  )
+
+  def __init__(self) -> None:
+    super().__init__()
+
+    self.gyro = (
+      AHRS(
+        wpilib.SerialPort.Port.kUSB,
+        AHRS.SerialDataType.kProcessedData,
+        int(1 / RobotConstants.period),
+      )
+      if RobotConstants.navxPort == RobotConstants.NavXPort.kUSB
+      else AHRS(
+        wpilib.SPI.Port.kMXP,
+        int(1 / RobotConstants.period),
+      )
+    )
+
+    def resetGyro():
+      """reset gyro after it's calibration of 1s"""
+      sleep(1)
+      self.resetGyro()
+
+    Thread(target=resetGyro).start()
+
+    self.theta_pid = ProfiledPIDControllerRadians(
+      SwerveConstants.kPRobotTurn,
+      SwerveConstants.kIRobotTurn,
+      SwerveConstants.kDRobotTurn,
+      TrapezoidProfileRadians.Constraints(
+        SwerveConstants.kDriveMaxTurnMetersPerSecond,
+        SwerveConstants.kDriveMaxTurnAccelerationMetersPerSecond,
+      ),
+      period=RobotConstants.period,
+    )
+    self.theta_pid.enableContinuousInput(0, math.tau)
+    self.theta_pid.setTolerance(math.radians(3))
+
+    if not RobotConstants.isSimulation:
+      self.field = Field2d()
+      SmartDashboard.putData("Field", self.field)
+
+  def getAngle(self) -> float:
+    # return self.gyro.getAngle() % 360
+    # return self.gyro.getFusedHeading()
+    if RobotConstants.isSimulation:
+      from physics import PhysicsEngine
+
+      return PhysicsEngine.simGyro.getAngle()
+
+    return -self.gyro.getYaw()
+
+  def getRotation2d(self):
+    return Rotation2d.fromDegrees(self.getAngle())
+
+  def getPose(self) -> Pose2d:
+    return self.odometer.getPose()
+
+  def resetGyro(self):
+    # self.gyro.zeroYaw()
+    self.gyro.reset()
+
+  def reset_motor_positions(self):
+    self.front_left.resetEncoders()
+    self.front_right.resetEncoders()
+    self.back_left.resetEncoders()
+    self.back_right.resetEncoders()
+
+  def resetOdometer(self, pose: Pose2d = Pose2d()):
+    self.odometer.resetPosition(
+      self.getRotation2d(),
+      pose,
+      self.front_left.getPosition(),
+      self.front_right.getPosition(),
+      self.back_left.getPosition(),
+      self.back_right.getPosition(),
+    )
+
+  def periodic(self) -> None:
+    # TODO print gyro angle, robot pose on dashboard
+
+    if not RobotConstants.isSimulation:
```

Comment:
Replace this with the proper way of determining if the code is running in a simulation (described above in another comment).